lib/active_merchant/billing/gateways/firstdata_e4.rb

```diff
@@ -237,13 +236,21 @@ def add_credit_card(xml, credit_card, options)
           xml.tag! "CardHoldersName", credit_card.name
           xml.tag! "CardType", card_type(credit_card.brand)
 
-          eci = (credit_card.respond_to?(:eci) ? credit_card.eci : nil) || options[:eci] || DEFAULT_ECI
-          xml.tag! "Ecommerce_Flag", eci
-
+          add_eci(xml, credit_card, options)
           add_credit_card_verification_strings(xml, credit_card, options)
         end
       end
 
+      def add_eci(xml, credit_card, options)
+        eci = if credit_card.is_a?(NetworkTokenizationCreditCard) && credit_card.brand == "discover"
+          "04"
```

Comment:
IMO this is a failure of abstraction on FirstData's part. Who knows what other gateways do. I'd assume most normalize data across card networks, but here we're being bit by a leaky abstraction.